src/Storages/ExternalStream/Kafka/KafkaSink.h

```diff
@@ -56,42 +52,62 @@ class KafkaSink final : public SinkToStorage
     void checkpoint(CheckpointContextPtr) override;
 
 private:
-    static void onMessageDelivery(rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void * opaque)
-    {
-        static_cast<KafkaSink *>(opaque)->wb->onMessageDelivery(msg);
-    }
-
-    static int32_t onPartitioning(
-        const rd_kafka_topic_t * /*rkt*/,
-        const void * /*keydata*/,
-        size_t /*keylen*/,
-        int32_t partition_count,
-        void * rkt_opaque,
-        void * msg_opaque)
-    {
-        /// update partition count
-        auto * sink = static_cast<KafkaSink *>(rkt_opaque);
-        sink->partition_cnt = partition_count;
-
-        auto parition_id = static_cast<Int32>(reinterpret_cast<std::uintptr_t>(msg_opaque));
-        /// This should not really happen because Kafka does not support reducing partitions.
-        /// However, KIP-694 is currently under discussion, so this might heppen in the future.
-        if (parition_id >= partition_count)
-            parition_id = partition_count - 1;
-        return parition_id;
-    }
+    /// Callback for Kafka message delivery report
+    static void onMessageDelivery(rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void *  /*opaque*/);
+    void onMessageDelivery(const rd_kafka_message_t * msg);
+
+    void addMessageToBatch(char * pos, size_t len);
+
+    /// the number of acknowledgement has been received so far for the current checkpoint period
+    size_t acked() const { return state.acked; }
+    /// the number of errors has been received so far for the current checkpoint period
+    size_t error_count() const { return state.error_count; }
+    /// the number of outstanding messages for the current checkpoint period
+    size_t outstandings() const { return state.outstandings; }
+    /// the last error code received from delivery report callback
+    rd_kafka_resp_err_t lastSeenError() const { return static_cast<rd_kafka_resp_err_t>(state.last_error_code.load()); }
+    /// check if there are no more outstandings (i.e. delivery reports have been recieved
+    /// for all out-go messages, regardless if a message is successfully delivered or not)
+    bool hasNoOutstandings() const { return state.outstandings == state.acked + state.error_count; }
```

Comment:
Can we flip the logic here to make it more readable, like 

```
bool hasOutstandingMessages(...)
{
    return state.outstandings != state.acked + state.error_count;
}
```

Nit: renaming `outstanding()` to `outstandingMessages()` is probably more readable.